internal/controllers/crd_controller.go

```
@@ -0,0 +1,96 @@
+
+package controllers
+
+import (
+	"context"
+	_ "embed"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/internalversion/scheme"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const policyendpointsCrdName = "policyendpoints.networking.k8s.aws"
+
+//go:embed crds.yaml
+var policyEndpointsCrd string
+
+func decodePolicyEndpointsCrd() (*apiextensionsv1.CustomResourceDefinition, error) {
+	decoder := scheme.Codecs.UniversalDeserializer()
+	obj := &apiextensionsv1.CustomResourceDefinition{}
+
+	_, _, err := decoder.Decode([]byte(policyEndpointsCrd), nil, obj)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode CRD from embedded YAML: %w", err)
+	}
+	return obj, nil
+}
+
+func NewCRDReconciler(k8sClient client.Client, logger logr.Logger) *CRDReconciler {
+	return &CRDReconciler{
+		k8sClient: k8sClient,
+		logger:    logger,
+	}
+}
+
+// CRDReconciler reconciles a CRD object
+type CRDReconciler struct {
+	k8sClient client.Client
+	logger    logr.Logger
+}
+
+// +kubebuilder:rbac:groups=extensions,resources=crds,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=extensions,resources=crds/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=extensions,resources=crds/finalizers,verbs=update
+// +kubebuilder:rbac:groups=apiextensions.k8s.io,resources=customresourcedefinitions,verbs=get;list;watch;create;update
+
+func (r *CRDReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = logf.FromContext(ctx)
+	r.logger.Info("Got reconcile request", "resource", req)
+	if req.Name != policyendpointsCrdName {
+		r.logger.Info("Ignoring reconcile request for non-policyendpoints CRD", "name", req.Name)
+		return ctrl.Result{}, nil
+	}
+	desiredCRD, err := decodePolicyEndpointsCrd()
```

do we need to decode this for every `Reconcile()` call? or we can do once while creating CRDReconciler object?

looks like a static one time thing. we can probably fail fast if `crds.yaml` is not found. Also i think we can reference `policyendpointsCrdName` after decoding the CRD only.

---

will update to skip this every time